presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java

```diff
@@ -156,6 +156,7 @@ public void addPage(Page page, int partitionId, Block partitionIds)
         for (int i = 0; i < channels.length; i++) {
             Block block = page.getBlock(i);
             channels[i].add(block);
+            pagesMemorySize += block.getRetainedSizeInBytes() / partitionCount;
```

Comment:
To be honest I don't like this way of memory accounting :( I know that I had did the same way, but maybe we indeed need those shared memory reservation with reference counting? It seems much cleaner to me and less prone to errors.